core/src/main/java/com/linecorp/armeria/client/endpoint/WeightBasedRandomEndpointSelector.java

```diff
@@ -0,0 +1,148 @@
+ * under the License.
+ */
+package com.linecorp.armeria.client.endpoint;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.annotation.Nullable;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.client.Endpoint;
+
+final class WeightBasedRandomEndpointSelector {
+
+    private final List<Entry> entries;
+    private final long totalWeight;
+    private final List<Entry> currentEntries;
+    private long currentTotalWeight;
+
+    WeightBasedRandomEndpointSelector(List<Endpoint> endpoints) {
+        final ImmutableList.Builder<Entry> builder = ImmutableList.builder();
+
+        long totalWeight = 0;
+        for (Endpoint endpoint : endpoints) {
+            builder.add(new Entry(endpoint, totalWeight));
+            totalWeight += endpoint.weight();
+        }
+        this.totalWeight = totalWeight;
+        currentTotalWeight = totalWeight;
+        entries = builder.build();
+        currentEntries = new CopyOnWriteArrayList<>(entries);
+    }
+
+    @VisibleForTesting
+    List<Entry> entries() {
+        return entries;
+    }
+
+    @Nullable
+    Endpoint selectEndpoint() {
+        if (entries.isEmpty()) {
+            return null;
+        }
+        for (;;) {
+            // selectEndpoint0() returns null if the endpoint is selected more than its weight while
+            // other endpoints are not. In that case the entry of that endpoint is removed and returns null.
+            // So we should loop to select another endpoint.
+            // This guarantees that the endpoint whose weight is very low is always selected when
+            // selectEndpoint() is called by totalWeight times even though we use random.
+            final Endpoint endpoint = selectEndpoint0();
+            if (endpoint != null) {
+                return endpoint;
+            }
+        }
+    }
+
+    @Nullable
+    Endpoint selectEndpoint0() {
+        final long nextLong = ThreadLocalRandom.current().nextLong(currentTotalWeight);
+        // There's a chance that currentTotalWeight is changed before looping currentEntries.
+        // However, we have counters and choosing an endpoint doesn't have to be exact so no big deal.
+        // TODO(minwoox): Use binary search when the number of endpoints is greater than N.
+        Endpoint selected = null;
+        for (Entry entry : currentEntries) {
+            if (entry.lowerBound >= nextLong) {
+                if (entry.increaseCounter()) {
+                    selected = entry.endpoint();
+                }
+                if (!entry.isFull()) {
+                    return selected;
+                }
+
+                // The entry is full so we should remove the entry from currentEntries.
+                synchronized (currentEntries) {
```

Comment:
Just wondering, is the removal easier to reason about than making entries as unavailable? Something like this?

```java
if (!entry.isFull()) {
  return selected;
} else if (entry.isUnavailable()) {
  continue;
}
entry.markUnavailable();
currentTotalWeight.decrementAndGet(entry.endpoint().weight());
```

Reply:
Well I figured an unavailable marker should be equivalent to removal, just without having to manage the list and maybe easier to mutate it atomically. `if (unavailable) continue;` feels equivalent semantically to removal. But maybe missing some corner case.

Reply:
I now realized the logic I implemented was not working as I expected. 😅  Maybe that's the reason that I got confused.
The current logic is slanted because the entry, which is next to the endpoint whose weight is large, will get much chance to get selected after the endpoint is full.
So I have to recalculate the `lowerBound` after the entry is removed but I forgot about it. 😅 
On the other hand, I'm a bit doubtful about recalculation. Let me think about this. 🤔

Reply:
@anuraaga I updated the logic a bit. PTAL. 🙇‍♂️